### PR TITLE
`verdi calcjob gotocomputer`: Add proxy command

### PR DESCRIPTION
**Pull request by mbercx**
Currently the `verdi calcjob gocomputer` command does not work in case
the connection to the remote computer is via a proxy. Here we add the
`ProxyCommand` option to the parameters for the `ssh` command.

### CODE REVIEW

**Comment by mbercx** [comment]
I basically got sick enough of this not working on Piz Daint that I finally got around to fixing it. 😅

**Review by ltalirz — COMMENTED** [review]
thanks @mbercx , I ran into this in https://github.com/aiidateam/aiida-core/issues/4100#issuecomment-762325711 as well and it's good to fix.

Just to be sure - this change will make sure to read the proxycommand from the profile as well, is that correct?

Happy to do a proper review once the tests pass

**Comment by ltalirz** [comment]
just to mention a related issue that might make sense to fix in one go as your touching this bit of the code https://github.com/aiidateam/aiida-core/issues/3311

(optional, not required of course)

**Comment by mbercx** [comment]
@ltalirz tests are still failing because of #4860, but I think I found the issue that was causing them to fail previously. It was a little subtle. In the `open` method of the `SshTransport`, the `'key_filename'` and `'proxy_command'` keys are `pop`-ed from the `connection_arguments`:

https://github.com/aiidateam/aiida-core/blob/5ab86cca5fc2070937cfa90b1a479d06878e68dd/aiida/transports/plugins/ssh.py#L430-L432

This makes sense, since this variable is later passed to the `connect` command of the `SshClient`, which doesn't accept e.g. `proxy_command` as a keyword argument. However, this also mutated the `SshTransport._connect_args` dictionary, removing the `proxy_command`. So, I decided to `copy` this dictionary instead.

Re #3311: Now that I've had to explore the transport code a bit, I'd be happy to work on this as well, but would fix this in a separate PR since it's not yet clear what the best solution is and although it is _related_ to the transport code, the required changes are still quite different from those made in this PR. :)

**Comment by mbercx** [comment]
> Just to be sure - this change will make sure to read the proxycommand from the profile as well, is that correct?

I'm not sure, actually. You mean the default set in the `.ssh/config` file? I'm still trying to figure out how and where exactly this is read. Frankly, the transport code is a little convoluted. 😅 

The `gotocomputer_command` reads all the connection arguments from the `_connect_args` dictionary, which is populated based on the `kwargs` in the constructor of the `SshTransport` class. However, after doing a bit of digging, it seems that the defaults, parsed from the `.ssh/config` file in the `parse_sshconfig` method, are only obtained when the `transport_option_default` in the `transports/cli.py` module is called. Which I think is done during the creation of the options of the `configure` commands. So I'm assuming that when the computer is first configured, AiiDA does read these config settings from the `.ssh/config` file and add them to the `authinfo` of the computer?

**Review by ltalirz — APPROVED** [review]
thanks @mbercx , this looks good to me

**Comment by ltalirz** [comment]
Hi @mbercx - yes, when setting up a computer, AiiDA can parse the `~.ssh/config` file and store this information in the database.

> Just to be sure - this change will make sure to read the proxycommand from the profile as well, is that correct?

What I meant here is:
Different from the regular ssh transport (which uses `paramiko`), the `gotocomputer` command used to directly run `ssh` through bash, which results in it using the `~

> > Just to be sure - this change will make sure to read the proxycommand from the profile as well, is that correct?
> 
> I'm not sure, actually. You mean the default set in the `.ssh/config` file? I'm still trying to figure out how and where exactly this is read. Frankly, the transport code is a little convoluted. 😅

No, I meant that the `gotocomputer` command would now take the configuration from the database (which is indeed what you added).
the reason it *also* reads from the `~.ssh/config` file is that the `gotocomputer` command directly executes `ssh` on the command line (it has to), which then parses the `~/.ssh/config` file. This does not happen when going through `paramiko`.


> The `gotocomputer_command` reads all the connection arguments from the `_connect_args` dictionary, which is populated based on the `kwargs` in the constructor of the `SshTransport` class. However, after doing a bit of digging, it seems that the defaults, parsed from the `.ssh/config` file in the `parse_sshconfig` method, are only obtained when the `transport_option_default` in the `transports/cli.py` module is called. Which I think is done during the creation of the options of the `configure` commands. So I'm assuming that when the computer is first configured, AiiDA does read these config settings from the `.ssh/config` file and add them to the `authinfo` of the computer?

When you set up the computer, you can select whether AiiDA should parse the ssh/config or not.

**Comment by ltalirz** [comment]
I leave the merge to you.
The commit message should mention that this is a slight change in behavior: Instead of the proxycommand stored in `~.ssh/config`,  the `gotocomputer` command will now use the proxycommand stored in the database (which should anyhow work as well since it is used by AiiDA).

**Comment by mbercx** [comment]
> The reason it _also_ reads from the `~.ssh/config` file is that the `gotocomputer` command directly executes `ssh` on the command line (it has to), which then parses the `~/.ssh/config` file. This does not happen in other contexts when going through `paramiko`.

But, this would only work in case the user has set the `Host` as the `HostName`, right? I usually use shorthands for the `host`, e.g.:

```console
Host ela
  HostName ela.cscs.ch
  User mbercx
  IdentityFile ~/.ssh/id_rsa-daint
Host daint
  HostName daint.cscs.ch
  User mbercx
  IdentityFile ~/.ssh/id_rsa-daint
  ProxyJump ela
```

(Which is [in line with our documentation](https://aiida.readthedocs.io/projects/aiida-core/en/latest/howto/ssh.html#ssh-configuration)). But AiiDA requests the "fully qualified" host name:

```console
  -H, --hostname HOSTNAME         The fully qualified hostname of the computer
                                  (e.g. daint.cscs.ch). Use "localhost" when
                                  setting up the computer that AiiDA is
                                  running on.  [required]
```

So this is of course why `verdi gotocomputer` was failing in my case, because when you execute `ssh -t daint.cscs.ch ...`, my `.ssh/config` is not being parsed.

> When you set up the computer, you can select whether AiiDA should parse the ssh/config or not.

This faces the same problem. And of course `ProxyJump` is ignored, since paramiko doesn't support it.

PS: Actually, after doing a bit of digging while writing this comment, my whole explanation is raised in [this issue](https://github.com/aiidateam/aiida-core/issues/4100). 😅  I'll continue the discussion there!

**Comment by mbercx** [comment]
> I leave the merge to you.

Thanks for the review, Leo! Will merge once #4860 is merged.

**Comment by ltalirz** [comment]
thanks, will unsubscribe from this pr
feel free to merge without further review (e.g. merging in the postgres updates)

**Review by ramirezfranciscof — APPROVED** [review]
Already approved by @ltalirz , just leaving the formal thumbs up.

**Pull request merged**
